lldb/source/Target/Target.cpp

```diff
@@ -1568,15 +1569,16 @@ bool Target::SetArchitecture(const ArchSpec &arch_spec, bool set_platform,
 
       if (m_arch.GetSpec().IsCompatibleMatch(other)) {
         compatible_local_arch = true;
-        bool arch_changed, vendor_changed, os_changed, os_ver_changed,
-            env_changed;
 
-        m_arch.GetSpec().PiecewiseTripleCompare(other, arch_changed,
-                                                vendor_changed, os_changed,
-                                                os_ver_changed, env_changed);
-
-        if (!arch_changed && !vendor_changed && !os_changed && !env_changed)
+        if (m_arch.GetSpec().GetTriple() == other.GetTriple())
           replace_local_arch = false;
+        // Workaround for for pre-2024 debugserver, which always
+        // returns arm64e on arm64e-capable hardware regardless of
+        // what the process is. This can be deleted at some point in
+        // the future.
```

Comment:
@jasonmolenda I really wish we could report a version number from debugserver so we don't have to make guesses as to what's supported and what not. This is the second time this has come up in a few months (I think previously this came up in the context of large watchpoints?). Maybe it's time to bite the bullet?

Reply:
The packet is poorly defined to return a "version" which lldb expects to be able to convert to an integer, so what we get today with an Xcode debugserver:

```
(lldb) process plugin packet send qGDBServerVersion
  packet: qGDBServerVersion
response: name:debugserver;version:1500;
```

With github main we'll get the llvm version, so `name:debugserver;version:19;`, lol.

This is one of those "version has never been properly/really defined" so we could add a "vendor" (llvm.org, apple, linode, etc) and a version *string* that might contain more than the major number, the vendor string informing how to interpret the version.

Reply:
Another way of saying this -- we DO have a version packet, but it is kinda not doing what we need, so maybe we should redefine what it returns.  We added it but never actually used it in lldb, we can change what it returns without breaking anything.

Reply:
We face the exact same issue with the compiler and pretty much all other llvm tools that follow that versioning scheme. In other words I don't think we should go out of our way to fix that just for debugserver. A simple heuristic could be version < 1000 is llvm version and > 1000 is the Apple version and possibly have a table as those releases aren't necessary aligned.